core/go/internal/groupmgr/message_listener.go

```diff
@@ -0,0 +1,705 @@
+	l.done = make(chan struct{})
+}
+
+func (l *messageListener) start() {
+	if l.done == nil {
+		l.initStart()
+		go l.runListener()
+	}
+}
+
+func (l *messageListener) stop() {
+	if l.done != nil {
+		l.cancelCtx()
+		<-l.done
+		l.done = nil
+	}
+}
+
+func (l *messageListener) notifyNewMessages() {
+	select {
+	case l.newMessages <- true:
+	default:
+	}
+}
+
+func (l *messageListener) addReceiver(r components.PrivacyGroupMessageReceiver) *registeredMessageReceiver {
+	l.receiverLock.Lock()
+	defer l.receiverLock.Unlock()
+
+	registered := &registeredMessageReceiver{
+		id:                          uuid.New(),
+		l:                           l,
+		PrivacyGroupMessageReceiver: r,
+	}
+	l.receivers = append(l.receivers, registered)
+
+	select {
+	case l.newReceivers <- true:
+	default:
+	}
+
+	return registered
+}
+
+func (l *messageListener) removeReceiver(rid uuid.UUID) {
+	l.receiverLock.Lock()
+	defer l.receiverLock.Unlock()
+
+	if len(l.receivers) > 0 {
+		newReceivers := make([]*registeredMessageReceiver, 0, len(l.receivers)-1)
+		for _, existing := range l.receivers {
+			if existing.id != rid {
+				newReceivers = append(newReceivers, existing)
+			}
+		}
+		l.receivers = newReceivers
+	}
+}
+
+func (l *messageListener) loadCheckpoint() error {
+	var checkpoints []*persistedMessageCheckpoint
+	err := l.gm.p.DB().
+		WithContext(l.ctx).
+		Where("listener = ?", l.spec.Name).
+		Limit(1).
+		Find(&checkpoints).
+		Error
+	if err != nil {
+		return err
+	}
+	if len(checkpoints) == 0 {
+		if l.spec.Filters.SequenceAbove != nil {
+			l.checkpoint = l.spec.Filters.SequenceAbove
+			log.L(l.ctx).Infof("Started message listener with minSequence=%d", *l.checkpoint)
+		} else {
+			log.L(l.ctx).Infof("Started message listener from sequence 0")
+		}
+	} else {
+		cpSequence := checkpoints[0].Sequence
+		l.checkpoint = &cpSequence
+		log.L(l.ctx).Infof("Started message listener with checkpoint=%d", cpSequence)
+	}
+	return nil
+}
+
+func (l *messageListener) readPage() ([]*persistedMessage, error) {
+	var messages []*persistedMessage
+	err := l.gm.messagesRetry.Do(l.ctx, func(attempt int) (retryable bool, err error) {
+		db := l.gm.p.DB()
+		q := l.gm.buildListenerDBQuery(l.spec, db)
+		if l.checkpoint != nil {
+			q = q.Where(`"pgroup_msgs"."local_seq" > ?`, *l.checkpoint)
+		}
+		return true, q.Find(&messages).Error
+	})
+	return messages, err
+}
+
+func (l *messageListener) processPersistedMessage(b *messageDeliveryBatch, pm *persistedMessage) {
+	if !l.checkMatch(pm) {
+		return
+	}
+	// Otherwise we can process the message
+	log.L(l.ctx).Infof("Added message %d/%s (domain='%s') to batch %d", pm.LocalSeq, pm.ID, pm.Domain, b.ID)
+	b.Messages = append(b.Messages, pm.mapToAPI())
+}
+
+func (l *messageListener) nextReceiver(b *messageDeliveryBatch) (r components.PrivacyGroupMessageReceiver, err error) {
+
+	for {
+		l.receiverLock.Lock()
+		if len(l.receivers) > 0 {
+			r = l.receivers[int(b.ID)%len(l.receivers)]
```

Comment:
Is this round-robin delivery?

Reply:
Yes, of batches (and associated acknowledgement), and only when you have multiple receivers attached to a single listener (multiple active subscribed websockets).